clap/codepoint_counting_writer.zig

```diff
@@ -1,101 +1,133 @@
-/// A Writer that counts how many codepoints has been written to it.
-/// Expects valid UTF-8 input, and does not validate the input.
-pub fn CodepointCountingWriter(comptime WriterType: type) type {
-    return struct {
-        codepoints_written: u64,
-        child_stream: WriterType,
-
-        pub const Error = WriterType.Error || error{Utf8InvalidStartByte};
-        pub const Writer = std.io.Writer(*Self, Error, write);
-
-        const Self = @This();
-
-        pub fn write(self: *Self, bytes: []const u8) Error!usize {
-            const bytes_and_codepoints = try utf8CountCodepointsAllowTruncate(bytes);
-            // Might not be the full input, so the leftover bytes are written on the next call.
-            const bytes_to_write = bytes[0..bytes_and_codepoints.bytes];
-            const amt = try self.child_stream.write(bytes_to_write);
-            const bytes_written = bytes_to_write[0..amt];
-            self.codepoints_written += (try utf8CountCodepointsAllowTruncate(bytes_written)).codepoints;
-            return amt;
-        }
+const std = @import("std");
+const testing = std.testing;
 
-        pub fn writer(self: *Self) Writer {
-            return .{ .context = self };
+/// A Writer that counts the number of UTF-8 codepoints written to it,
+/// while forwarding all bytes to a child writer.
+/// This implementation correctly handles codepoints split across buffer boundaries.
+pub const CodepointCountingWriter = struct {
+    child_writer: *std.Io.Writer,
+    codepoints_written: u64 = 0,
+
+    // Buffer to hold partial UTF-8 sequences between drain calls.
+    partial_buf: [4]u8 = .{0} ** 4,
+    partial_len: u2 = 0,
+
+    interface: std.Io.Writer,
+
+    const Self = @This();
+
+    pub fn init(child_writer: *std.Io.Writer, buffer: []u8) Self {
+        return .{
+            .child_writer = child_writer,
+            .interface = .{
+                .buffer = buffer,
+                .vtable = &.{ .drain = drain },
+            },
+        };
+    }
```

Comment:
As I've commented on before, I'd say just get rid of the buffering for this writer. I'd say, nearly always the underlying writer should be buffered and not this one.

If there are performance benefits to having a buffer on this writer, I'd like to see some measurements. Otherwise it seems to just complicate the `drain` function